src/tools/rbd/action/MirrorPool.cc

```diff
@@ -110,39 +145,180 @@ int get_remote_cluster_spec(const po::variables_map &vm,
+              << cpp_strerror(r) << std::endl;
+    return r;
+  }
+
+  bool json_valid = false;
+  json_spirit::mValue json_root;
+  if(json_spirit::read(out_bl.to_str(), json_root)) {
+    try {
+      auto& json_obj = json_root.get_obj();
+      for (auto& pairs : json_obj) {
+        (*attributes)[pairs.first] = pairs.second.get_str();
+      }
+      json_valid = true;
+    } catch (std::runtime_error&) {
+    }
+  }
+
+  if (!json_valid) {
+    std::cerr << "rbd: corrupt mirroring peer config" << std::endl;
+    return -EINVAL;
+  }
+  return 0;
+}
+
+int update_peer_config_key(librados::Rados& rados, int64_t pool_id,
+                           const std::string& peer_uuid,
+                           const std::string& key,
+                           const std::string& value) {
+  std::map<std::string, std::string> attributes;
+  int r = get_peer_config_key(rados, pool_id, peer_uuid, &attributes);
+  if (r == -ENOENT) {
+    return set_peer_config_key(rados, pool_id, peer_uuid, {{key, value}});
+  } else if (r < 0) {
+    return r;
+  }
+
+  if (value.empty()) {
+    attributes.erase(key);
+  } else {
+    attributes[key] = value;
+  }
+  return set_peer_config_key(rados, pool_id, peer_uuid, std::move(attributes));
+}
+
+int format_mirror_peers(librados::Rados& rados, int64_t pool_id,
+                        at::Format::Formatter formatter,
+                        const std::vector<librbd::mirror_peer_t> &peers,
+                        bool config_key) {
+  TextTable tbl;
   if (formatter != nullptr) {
     formatter->open_array_section("peers");
-    for (auto &peer : peers) {
-      formatter->open_object_section("peer");
-      formatter->dump_string("uuid", peer.uuid);
-      formatter->dump_string("cluster_name", peer.cluster_name);
-      formatter->dump_string("client_name", peer.client_name);
-      formatter->close_section();
-    }
-    formatter->close_section();
   } else {
     std::cout << "Peers: ";
     if (peers.empty()) {
       std::cout << "none" << std::endl;
     } else {
-      TextTable tbl;
       tbl.define_column("", TextTable::LEFT, TextTable::LEFT);
       tbl.define_column("UUID", TextTable::LEFT, TextTable::LEFT);
       tbl.define_column("NAME", TextTable::LEFT, TextTable::LEFT);
       tbl.define_column("CLIENT", TextTable::LEFT, TextTable::LEFT);
-      for (auto &peer : peers) {
-        tbl << " "
-            << peer.uuid
-            << peer.cluster_name
-            << peer.client_name
-            << TextTable::endrow;
+      if (config_key) {
+        tbl.define_column("MON_HOST", TextTable::LEFT, TextTable::LEFT);
+        tbl.define_column("KEY", TextTable::LEFT, TextTable::LEFT);
+      }
+    }
+  }
+
+  for (auto &peer : peers) {
+    std::map<std::string, std::string> attributes;
+    if (config_key) {
+      int r = get_peer_config_key(rados, pool_id, peer.uuid, &attributes);
+      if (r < 0 && r != -ENOENT) {
+        return r;
```

Comment:
I was originally thinking that if you specified the ```--all`` optional and we had issues, we should fail since we cannot satisfy the user's request. I am on the fence about adding a warning for ```-EPERM```, but I could add it.

Reply:
I was thinking that we could output something like "<hidden>" or "<forbidden>" (can't think out a good message), so the user would be aware it got not full info. But not sure then about exit status, should it be 0 or not.

Actually I don't have a strong opinion here and I am fine with the current version. Or just a more meaningful error message for `-EPERM` could make it more user friendly. I just wanted to be sure you were aware about this.

Reply:
I like that idea, I'll add a meaningful error message for the ```-EPERM``` case

Reply:
Updated ```get_peer_config_key``` to output a nicer error message when it gets an ```-EPERM``` failure

Reply:
I hit the same issue as well during my local testing where it worked because I set the variable and then it failed on teuthology.